File: articles/research-methods-in-developmental-psychology-longitudinal-cross-sectional-and-cohort-designs/go/design_summary.go
package main

import (
	"encoding/csv"
	"fmt"
	"log"
	"os"
	"strconv"
)

func findColumn(header []string, name string) int {
	for i, col := range header {
		if col == name {
			return i
		}
	}
	return -1
}

func main() {
	if len(os.Args) < 2 {
		log.Fatal("Usage: go run design_summary.go data/developmental_design_panel.csv")
	}

	file, err := os.Open(os.Args[1])
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	reader := csv.NewReader(file)
	rows, err := reader.ReadAll()
	if err != nil {
		log.Fatal(err)
	}

	if len(rows) < 2 {
		log.Fatal("CSV has no data rows")
	}

	header := rows[0]
	scoreIdx := findColumn(header, "development_score")
	observedIdx := findColumn(header, "observed")
	ageIdx := findColumn(header, "age")

	if scoreIdx < 0 || observedIdx < 0 || ageIdx < 0 {
		log.Fatal("Required columns not found")
	}

	count := 0
	scoreSum := 0.0
	ageSum := 0.0

	for _, row := range rows[1:] {
		observed, err := strconv.Atoi(row[observedIdx])
		if err != nil || observed != 1 {
			continue
		}

		score, err1 := strconv.ParseFloat(row[scoreIdx], 64)
		age, err2 := strconv.ParseFloat(row[ageIdx], 64)

		if err1 != nil || err2 != nil {
			continue
		}

		count++
		scoreSum += score
		ageSum += age
	}

	if count == 0 {
		log.Fatal("No observed rows found")
	}

	fmt.Printf("Observed rows analyzed: %d\n", count)
	fmt.Printf("Mean age: %.4f\n", ageSum/float64(count))
	fmt.Printf("Mean development_score: %.4f\n", scoreSum/float64(count))
}

File: articles/research-methods-in-developmental-psychology-longitudinal-cross-sectional-and-cohort-designs/cpp/design_summary.cpp
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

std::vector<std::string> split_csv_line(const std::string& line) {
    std::vector<std::string> values;
    std::stringstream ss(line);
    std::string item;

    while (std::getline(ss, item, ',')) {
        values.push_back(item);
    }

    return values;
}

int find_column(const std::vector<std::string>& header, const std::string& name) {
    for (size_t i = 0; i < header.size(); ++i) {
        if (header[i] == name) {
            return static_cast<int>(i);
        }
    }
    return -1;
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        std::cerr << "Usage: design_summary data/developmental_design_panel.csv\n";
        return 1;
    }

    std::ifstream file(argv[1]);

    if (!file.is_open()) {
        std::cerr << "Unable to open CSV file\n";
        return 1;
    }

    std::string line;
    std::getline(file, line);

    auto header = split_csv_line(line);
    int score_col = find_column(header, "development_score");
    int observed_col = find_column(header, "observed");
    int age_col = find_column(header, "age");

    if (score_col < 0 || observed_col < 0 || age_col < 0) {
        std::cerr << "Required columns not found\n";
        return 1;
    }

    long count = 0;
    double score_sum = 0.0;
    double age_sum = 0.0;

    while (std::getline(file, line)) {
        auto values = split_csv_line(line);

        if (
            static_cast<int>(values.size()) <= score_col ||
            static_cast<int>(values.size()) <= observed_col ||
            static_cast<int>(values.size()) <= age_col
        ) {
            continue;
        }

        int observed = std::stoi(values[observed_col]);
        if (observed != 1) {
            continue;
        }

        score_sum += std::stod(values[score_col]);
        age_sum += std::stod(values[age_col]);
        count += 1;
    }

    if (count == 0) {
        std::cerr << "No observed rows found\n";
        return 1;
    }

    std::cout << "Observed rows analyzed: " << count << "\n";
    std::cout << "Mean age: " << age_sum / count << "\n";
    std::cout << "Mean development_score: " << score_sum / count << "\n";

    return 0;
}
